### PR TITLE
Cache legend images

### PR DESCRIPTION
**Pull request by chrismayer**
This caches the legend images as dataURLs in an internal lookup, so the images only have to be queried at the first activation of the legend graphic in the tree.

@geographika Could you please give it a try in an application with a large amount of layers in order to test if this brings a performance boost for the layer tree? TIA! 

Relates to #274

### CODE REVIEW

**Comment by geographika** [comment]
@chrismayer - is there a new file missing `CpsiMapview.util.Html`? Getting xtemplate errors at the moment.

**Comment by chrismayer** [comment]
Sorry you are right...added missing file with e56eb63

**Comment by geographika** [comment]
@chrismayer - it does seem to refresh a little quicker - and cuts down on lots of `GetLegendGraphic` requests. 
The cache is a nice idea. There may be a requirement for "data-driven" legends or scale dependent legends in the future, but those could probably be excluded with a config setting. 

In regard to overall performance it appears the entire tree is redrawn whenever any layer changes visibility - any idea if this is required?
Could this full redraw be caused by https://github.com/geoext/geoext3/pull/62 - `toggleParentNodes` seems to loop through every layer setting properties on each model:

```
                    parent.set('__toggleMode', 'ol3'); // prevents recursion
                    parent.set('checked', true);
                    parent.set('__toggleMode', 'classic');
```

Could it be this causing the full redraw?
I recently had an issue with models having properties set to `dirty` when not required which can be avoided with an options object and `dirty` property - see [set docs](https://docs.sencha.com/extjs/6.7.0/classic/Ext.data.Model.html#method-set). 
However, this may be required to keep children/parent checkboxes in synch. 

```
parent.set('__toggleMode', 'classic', {dirty: false});
```

I tried with the alternative [ol3 toggle mode](https://geoext.github.io/geoext3/master/docs/#!/api/GeoExt.data.store.LayersTree-cfg-folderToggleMode) but it still seems to redraw all nodes on the tree.

File: app/util/Html.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+         * Add events to HTML elements in a cross browser compatible manor.
+         *
+         * Credits: http://stackoverflow.com/a/21140463
+         *
+         * @param {DOMElement} htmlElement The element to attach the event to.
+         * @param {String} eventName The name of the event to attach.
+         * @param {Function} eventFunction The handler for the `eventName` on
+         *     the passed `htmlElement`.
+         */
+        addEvent: function (htmlElement, eventName, eventFunction) {
```
Comment (geographika) [inline app/util/Html.js:17]: I presume ExtJS doesn't already have a function for this that could be used?

File: app/util/Legend.js [hunk 2]
```diff
@@ -150,5 +151,63 @@ Ext.define('CpsiMapview.util.Legend', {
...
+                canvas.height = height;
+
+                var ctx = canvas.getContext('2d');
+                ctx.drawImage(this, 0, 0);
+                var legendDataUrl = canvas.toDataURL('image/png');
+
+                // store in cache for later reuse
+                CpsiMapview.view.LayerTree.legendImgLookup[layerKey] = legendDataUrl;
+
+                LegendUtil['legendLoading_' + layerKey] = false;
```
Comment (geographika) [inline app/util/Legend.js:205]: Note to self: this is where the isLoading flag is set to false

**Review by geographika — APPROVED** [review]

File: app/view/LayerTree.js [hunk 3]
```diff
@@ -18,6 +18,10 @@ Ext.define('CpsiMapview.view.LayerTree', {
 
     controller: 'cmv_layertree',
 
+    statics: {
```
[the remark this answers is not in the document]
Comment (chrismayer) [reply app/view/LayerTree.js:21]: Yes, this was the idea and the reference it quite easier. But I can check if I bind it to the LayerTree instance, if you prefer.

File: app/util/Html.js (hunk 1 above)
Comment (chrismayer) [reply app/util/Html.js:17]: As far as I know not, but I'll double check.

**Comment by chrismayer** [comment]
Thanks for the feedback @geographika. I'll suggest to switch the general LayerTree performance discussion to #274 since this here is only only aspect regarding to optimize performance.

File: app/view/LayerTree.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (geographika) [reply app/view/LayerTree.js:21]: No need for any changes it was just for my understanding

File: app/util/Html.js (hunk 1 above)
Comment (geographika) [reply app/util/Html.js:17]: Guess if there was it would be part of https://docs.sencha.com/extjs/6.2.0/classic/Ext.dom.Element.html ?

**Comment by geographika** [comment]
@chrismayer - is this ready to merge? It stops lots of unnecessary requests so helps speed up the application. 
The function has access to the layer so if there were scale-dependent legends it looks like it should be simple enough to disable caching with an option.

File: app/util/Html.js (hunk 1 above)
Comment (chrismayer) [reply app/util/Html.js:17]: Probably yes. I had a quick look and it seems `Ext.Dom.Element` has the normal `on` event binding method. Maybe it is worth giving this a try so we could get rid of the HTML util class.

**Comment by chrismayer** [comment]
> @chrismayer - is this ready to merge?

Has some potential for optimization, like getting rid of the HTML util class (see comment above) or prevent more `GetLegend` requests at startup. But this can surely be done in follow-up PRs in a later stage. As already stated this should bring another performance boost, so I pledge to merge. Or what do you think @geographika ?

**Comment by geographika** [comment]
@chrismayer - agree!

**Pull request merged**